### PR TITLE
Same code path for dynamic mappings updates and updates coming from the API.

### LINKED ISSUE #9364: Mappings: dynamic mapping updates should use the same path as mapping updates from the API
Currently we have two different code paths for dynamic mapping updates and mapping updates that come from the API.

I think we should improve it by making document parsing generate mapping updates, and then apply these mapping updates just like we apply mapping updates that come from the API.

This would probably help with #8688

### PR DESCRIPTION
**Pull request by jpountz**
We have two completely different code paths for mappings updates, depending on
whether they come from the API or are guessed based on the parsed documents.
This commit makes dynamic mappings updates execute like updates from the API.

The only change in behaviour is that a document that fails parsing can not
modify mappings anymore (useful to prevent issues such as #9851). Other than
that, this change should be fairly transparent to users but working this way
opens doors to other changes such as validating dynamic mappings updates on the
master node (#8688).

The way it works internally is that Mapper.parse now returns a Mapper instead
of being void. The returned Mapper represents a mapping update that has been
performed in order to parse the document. Mappings updates are propagated
recursively back to the root mapper, and once parsing is finished, we check
that the mappings update can be applied, and either fail the parsing if the
update cannot be merged (eg. because of a concurrent mapping update from the
API) or merge the update into the mappings.

However not all mappings updates can be applied recursively, `copy_to` for
instance can add mappings at totally different places in the tree. Because of
it I added ParseContext.rootMapperUpdates which `copy_to` fills when the
field to copy data to does not exist in the mappings yet. These mappings
updates are merged from the ones generated by regular parsing.

One particular mapping update was the `auto_boost` setting on the `all` root
mapper. Being tricky to work on, I removed it in favour of search-time checks
that payloads have been indexed.

One interesting side-effect of the change is that concurrency on ObjectMapper
is greatly simplified since we do not have to care anymore about having
concurrent dynamic mappings and API updates.

Closes #9364

### CODE REVIEW

File: src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java [hunk 1]
```diff
@@ -637,8 +665,41 @@ public void traverse(ObjectMapperListener listener) {
         rootObjectMapper.traverse(listener);
     }
 
+    private MergeContext mergeContext(MergeFlags mergeFlags) {
```
Comment (rjernst) [inline src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java]: maybe rename to `newMergeContext` or `createMergeContext`?

File: src/main/java/org/elasticsearch/index/mapper/object/ObjectMapper.java [hunk 2]
```diff
@@ -966,33 +944,30 @@ public void merge(final Mapper mergeWith, final MergeContext mergeContext) throw
...
             }
         }
-
+        if (!newFieldMappers.mappers.isEmpty()) {
+            mergeContext.addFieldMappers(newFieldMappers.mappers);
+        }
+        if (!newObjectMappers.mappers.isEmpty()) {
+            mergeContext.addObjectMappers(newObjectMappers.mappers);
+        }
+        // and the mappers only after the administration have been done, so it will not be visible to parser (which first try to read with no lock)
```
Comment (rjernst) [inline src/main/java/org/elasticsearch/index/mapper/object/ObjectMapper.java]: and -> add?

File: src/main/java/org/elasticsearch/index/mapper/internal/AllFieldMapper.java [hunk 3]
```diff
@@ -155,8 +149,10 @@ public AllFieldMapper build(BuilderContext context) {
                     builder.enabled(nodeBooleanValue(fieldNode) ? EnabledAttributeMapper.ENABLED : EnabledAttributeMapper.DISABLED);
                     iterator.remove();
                 } else if (fieldName.equals("auto_boost")) {
-                    builder.autoBoost = nodeBooleanValue(fieldNode);
-                    iterator.remove();
+                    if (parserContext.indexVersionCreated().before(Version.V_2_0_0)) {
```
Comment (rjernst) [inline src/main/java/org/elasticsearch/index/mapper/internal/AllFieldMapper.java]: join this into the else if? this way for 2.0 auto_boost will be rejected as an unknown setting.

**Comment by rjernst** [comment]
@jpountz This is *awesome*! LGTM

**Comment by jpountz** [comment]
Thanks @rjernst !

**Pull request merged**